engine/execution/ingestion/stop_control.go

```diff
@@ -3,318 +3,609 @@ package ingestion
+	// desired StopHeight, the first value new version should be used,
 	// so this height WON'T be executed
-	stopHeight uint64
+	StopHeight uint64
+
+	// if the node should crash or just pause after reaching StopHeight
+	ShouldCrash bool
+}
+
+type stopBoundary struct {
+	StopParameters
 
-	// if the node should crash or just pause after reaching stopHeight
-	crash bool
+	// once the StopParameters are reached they cannot be changed
+	cannotBeChanged bool
 
 	// This is the block ID of the block that should be executed last.
 	stopAfterExecuting flow.Identifier
 
-	log   zerolog.Logger
-	state StopControlState
+	// if the stop parameters were set by the version beacon
+	fromVersionBeacon bool
+}
+
+// String returns string in the format "crash@20023[versionBeacon]" or
+// "crash@20023@blockID[versionBeacon]"
+// block ID is only present if stopAfterExecuting is set
+// the ID is from the block that should be executed last and has height one
+// less than StopHeight
+func (s *stopBoundary) String() string {
+	if s == nil {
+		return "none"
+	}
+
+	sb := strings.Builder{}
+	if s.ShouldCrash {
+		sb.WriteString("crash")
+	} else {
+		sb.WriteString("stop")
+	}
+	sb.WriteString("@")
+	sb.WriteString(fmt.Sprintf("%d", s.StopHeight))
+
+	if s.stopAfterExecuting != flow.ZeroID {
+		sb.WriteString("@")
+		sb.WriteString(s.stopAfterExecuting.String())
+	}
+	if s.fromVersionBeacon {
+		sb.WriteString("[versionBeacon]")
+	} else {
+		sb.WriteString("[manual]")
+	}
+
+	return sb.String()
+}
 
-	// used to prevent setting stopHeight to block which has already been executed
-	highestExecutingHeight uint64
+type StopControlOption func(*StopControl)
+
+func StopControlWithLogger(log zerolog.Logger) StopControlOption {
+	return func(s *StopControl) {
+		s.log = log.With().Str("component", "stop_control").Logger()
+	}
 }
 
-type StopControlState byte
-
-const (
-	// StopControlOff default state, envisioned to be used most of the time.
-	// Stopping module is simply off, blocks will be processed "as usual".
-	StopControlOff StopControlState = iota
-
-	// StopControlSet means stopHeight is set but not reached yet,
-	// and nothing related to stopping happened yet.
-	// We could still go back to StopControlOff or progress to StopControlCommenced.
-	StopControlSet
-
-	// StopControlCommenced indicates that stopping process has commenced
-	// and no parameters can be changed anymore.
-	// For example, blocks at or above stopHeight has been received,
-	// but finalization didn't reach stopHeight yet.
-	// It can only progress to StopControlPaused
-	StopControlCommenced
-
-	// StopControlPaused means EN has stopped processing blocks.
-	// It can happen by reaching the set stopping `stopHeight`, or
-	// if the node was started in pause mode.
-	// It is a final state and cannot be changed
-	StopControlPaused
-)
+func StopControlWithStopped() StopControlOption {
+	return func(s *StopControl) {
+		s.stopped = true
+	}
+}
+
+func StopControlWithVersionControl(
```

Comment:
should version beacons be set dynamically, as part of SetStop?

Reply:
Stop can be set using a version beacon (by defining a version boundary in the smart contract). For that to happen the stop control needs access to the indexed version beacons, and needs to be aware of its own version. That is what this is, a convenience to set that up. 

Its not necessary during (non-related) testing so I made it a constructor option.